src/Evolution/DgSubcell/Matrices.cpp

```diff
@@ -20,11 +20,76 @@
+  ASSERT(subcell_extents == Index<Dim>(subcell_extents[0]),
+         "The subcell mesh must be uniform but is " << subcell_extents);
+  switch (dg_mesh.quadrature(0)) {
+    case Spectral::Quadrature::GaussLobatto: {
+      static const auto cache_gl = make_static_cache<
+          CacheRange<
+              Spectral::minimum_number_of_points<
+                  Spectral::Basis::Legendre,
+                  Spectral::Quadrature::GaussLobatto>,
+              Spectral::maximum_number_of_points<Spectral::Basis::Legendre>>,
+          CacheRange<Spectral::minimum_number_of_points<
+                         Spectral::Basis::FiniteDifference,
+                         Spectral::Quadrature::CellCentered>,
+                     Spectral::maximum_number_of_points<
+                         Spectral::Basis::FiniteDifference>>>(
+          [](const size_t local_num_dg_points,
+             const size_t local_num_fd_points) {
+            return Spectral::interpolation_matrix<
+                Spectral::Basis::Legendre, Spectral::Quadrature::GaussLobatto>(
+                local_num_dg_points,
+                Spectral::collocation_points<
+                    Spectral::Basis::FiniteDifference,
+                    Spectral::Quadrature::CellCentered>(local_num_fd_points));
+          });
+      return cache_gl(dg_mesh.extents(0), subcell_extents[0]);
+    }
+    case Spectral::Quadrature::Gauss: {
+      static const auto cache_g = make_static_cache<
+          CacheRange<
+              Spectral::minimum_number_of_points<Spectral::Basis::Legendre,
+                                                 Spectral::Quadrature::Gauss>,
+              Spectral::maximum_number_of_points<Spectral::Basis::Legendre>>,
+          CacheRange<Spectral::minimum_number_of_points<
+                         Spectral::Basis::FiniteDifference,
+                         Spectral::Quadrature::CellCentered>,
+                     Spectral::maximum_number_of_points<
+                         Spectral::Basis::FiniteDifference>>>(
+          [](const size_t local_num_dg_points,
+             const size_t local_num_fd_points) {
+            return Spectral::interpolation_matrix<Spectral::Basis::Legendre,
+                                                  Spectral::Quadrature::Gauss>(
+                local_num_dg_points,
+                Spectral::collocation_points<
+                    Spectral::Basis::FiniteDifference,
+                    Spectral::Quadrature::CellCentered>(local_num_fd_points));
+          });
+      return cache_g(dg_mesh.extents(0), subcell_extents[0]);
+    }
+    default:
+      ERROR("Unsupported quadrature type in FD subcell projection matrix");
+  };
+}
+
+namespace {
 template <Spectral::Quadrature QuadratureType, size_t NumDgGridPoints1d,
           size_t Dim>
-Matrix projection_matrix_cache_impl_helper(const Index<Dim>& subcell_extents) {
+Matrix projection_matrix_3d(const Index<Dim>& subcell_extents) {
```

Comment:
Is this actually specific to 3D?  I don't see anything choosing functions based on dimension.

Reply:
ah, no. It's the "full dimensions". Do you have a good suggestion for the name? I was clearly focusing on the 3d case when I made the name change... I have a similar naming dilemma  for reconstruction. There's `DimByDim` but then for the full one do I do `AllDimensions`? Any suggestions are much appreciated!